python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py

```diff
@@ -146,16 +159,157 @@ def get_last_updated_timestamp(record: Optional[EventLogRecord]) -> Optional[flo
-                ),
+                check_spec.metadata and check_spec.metadata.get(FRESHNESS_PARAMS_METADATA_KEY),
                 f"Asset check {check_spec.key} didn't have expected metadata. Please ensure that the asset check is a freshness check.",
             )
+
+
+def get_last_update_time_lower_bound(
+    freshness_cron: Optional[str],
+    current_time: datetime.datetime,
+    lower_bound_delta: datetime.timedelta,
+    freshness_cron_timezone: str,
+) -> datetime.datetime:
+    """Get the lower bound of the last_update_time for the given cron schedule."""
+    if freshness_cron:
+        return (
+            check.not_none(
+                get_latest_completed_cron_tick(
+                    freshness_cron=freshness_cron,
+                    current_time=current_time,
+                    timezone=freshness_cron_timezone,
+                )
+            )
+            - lower_bound_delta
+        )
+    else:
+        return current_time - lower_bound_delta
+
+
+def get_latest_complete_partition_key(
+    freshness_cron: Optional[str],
+    current_time: datetime.datetime,
+    freshness_cron_timezone: str,
+    partitions_def: Optional[TimeWindowPartitionsDefinition],
+) -> Optional[str]:
+    """Get the latest complete partition key for the given cron schedule."""
+    if not partitions_def:
+        return None
+    latest_cron_tick_time = check.not_none(
+        get_latest_completed_cron_tick(
+            freshness_cron=freshness_cron,
+            current_time=current_time,
+            timezone=freshness_cron_timezone,
+        )
+    )
+    time_window = check.not_none(partitions_def.get_prev_partition_window(latest_cron_tick_time))
+    return partitions_def.get_partition_key_range_for_time_window(time_window).start
+
+
+def build_freshness_check(
+    asset_key: AssetKey,
+    freshness_cron: Optional[str],
+    freshness_cron_timezone: str,
+    severity: AssetCheckSeverity,
+    lower_bound_delta: datetime.timedelta,
+    asset_property_enforcement_lambda: Optional[Callable[[AssetsDefinition], bool]],
+) -> AssetChecksDefinition:
+    params_metadata = {}
+    if freshness_cron:
+        params_metadata[FRESHNESS_CRON_METADATA_KEY] = freshness_cron
+        params_metadata[FRESHNESS_CRON_TIMEZONE_METADATA_KEY] = freshness_cron_timezone
+    if lower_bound_delta:
+        params_metadata[LOWER_BOUND_DELTA_METADATA_KEY] = lower_bound_delta.total_seconds()
+
+    @asset_check(
+        asset=asset_key,
+        description="Evaluates freshness for targeted asset.",
+        name="freshness_check",
+        metadata={FRESHNESS_PARAMS_METADATA_KEY: params_metadata},
+    )
+    def the_check(context: AssetCheckExecutionContext) -> AssetCheckResult:
+        if asset_property_enforcement_lambda:
+            asset_property_enforcement_lambda(
+                context.job_def.asset_layer.asset_graph.get(asset_key).assets_def
+            )
+        current_time = pendulum.now(
+            tz=freshness_cron_timezone
```

Comment:
Btw this timezone only applies to the cron schedule, not the effective time, correct? Won't this cause bugs if the freshness timezone is different than the timezone of th entire system?

Reply:
I guess there are actually three timezones at play here.
- The timezone of the freshness cron.
- The timezone of the partitions definition
- The timezone of the system actually running the code.
I guess what we probably need here to only track a "freshness timestamp" in terms of the actual system time, and then translate that back and forth between the various timezones. I'll add a test for this pathology.

Reply:
Just to follow up, I added a test for the case of complex timezone interaction.